### PR TITLE
perf(statistics): rewrite for performance on very large communities

### PR DESCRIPTION
**Pull request by davwheat**
**Changes proposed in this pull request:**

The Statistics extension is currently a big issue on large communities, due to the way it's written.

Currently, all statistics data is loaded on the backend during page load, and bundled with the page's payload itself. On a community with about 1 million discussions, this case mean the admin dashboard takes over 10s to load. In addition, statistics data is loaded no matter which dashboard page is navigated to, causing unnecessary database queries and server load.

Here's our proposal:
- [x] Move statistics data out of initial payload, and into an XHR request
- [x] Split XHR request into lifetime and timed
- [x] Add new periods: previous 7 and 28 days
- [x] If needed, move widget to the extension page instead of dashboard
  - [ ] ~~prevents an XHR request on every admin dash visit~~
  - [ ] ~~could also instead have a button labelled "show statistics" which loads it on the main dash~~
  - [x] if we move, should add message stating as such so admins aren't confused, with link to ext page
  - [x] add partial stats widget to dashboard, with link to original
- [x] Cache stats server-side to prevent recalculation with multiple hits in succession

There are probably some other things we can do to also improve performance, but I feel this is a good starting point.

**Necessity**

- [x] Has the problem that is being solved here been clearly explained?
- [x] If applicable, have various options for solving this problem been considered?
- [x] For core PRs, does this need to be in core, or could it be in an extension?
- [x] Are we willing to maintain this for years / potentially forever?

**Confirmed**

- [x] Frontend changes: tested on a local Flarum installation.
- [x] Backend changes: tests are green (run `composer test`).
- [x] Core developer confirmed locally this works as intended.
- [x] Tests have been added, or are not appropriate here.

**Required changes:**

- [ ] Related documentation PR: (Remove if irrelevant)
- [ ] Related core extension PRs: (Remove if irrelevant)

### CODE REVIEW

File: extensions/statistics/less/admin.less [hunk 1]
```diff
@@ -84,11 +119,12 @@
   position: relative; /* for absolutely positioned tooltip */
 
   /* https://www.smashingmagazine.com/2015/11/using-system-ui-fonts-practical-guide/ */
-  font-family: -apple-system, BlinkMacSystemFont,
-  'Segoe UI', 'Roboto', 'Oxygen', 'Ubuntu', 'Cantarell',
-  'Fira Sans', 'Droid Sans', 'Helvetica Neue', sans-serif;
+  font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", "Roboto", "Oxygen",
+    "Ubuntu", "Cantarell", "Fira Sans", "Droid Sans", "Helvetica Neue",
+    sans-serif;
```
Comment (luceos) [inline extensions/statistics/less/admin.less]: I know this is nitpicking, but what happened here? Quotes? Shouldn't really be in this PR I think?

Comment (davwheat) [reply extensions/statistics/less/admin.less]: Oh, that was a side-effect of IDE code formatting. I can revert it.

Comment (davwheat) [reply extensions/statistics/less/admin.less]:
```suggestion
  font-family: -apple-system, BlinkMacSystemFont,
  'Segoe UI', 'Roboto', 'Oxygen', 'Ubuntu', 'Cantarell',
  'Fira Sans', 'Droid Sans', 'Helvetica Neue', sans-serif;
```

**Review by luceos — CHANGES_REQUESTED** [review]
I don't understand the need to use a setting for the timezone. And instead of the method returning the actual actor timezone, it returns the setting or the server timezone. I think only the frontend uses timezones right now, everything on the backend should always be uct. Can you clarify?

**Comment by davwheat** [comment]
In this PR, my main objective was improving performance.

The use of a setting, and returning the server timezone is just the original extension behaviour, which I didn't want to tamper with to reduce the chance of introducing bugs along with this PR.

I think it's easier for the PR to be reviewed based on the changes made compared to the current content and behaviour, and then continue to rewrite the behaviour in a second PR. I don't want this becoming one big huge PR which basically scraps the extension entirely and starts again.

That being said, I _believe_ the time zone things exist to account for the server not running on UTC, but I've not checked this.

**Review by luceos — APPROVED** [review]

File: extensions/statistics/locale/en.yml [hunk 2]
```diff
@@ -9,12 +9,16 @@ flarum-statistics:
 
     # These translations are used in the Statistics dashboard widget.
     statistics:
-      active_users_text: "{count} active"
```
Comment (SychO9) [inline extensions/statistics/locale/en.yml:12]: This is dead code I guess? can't see it used anywhere

File: extensions/statistics/src/Api/Controller/ShowStatisticsData.php [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+            return $cachedVal;
+        }
+
+        $val = array_map(function ($entity) {
+            return $entity[0]->count();
+        }, $this->entities);
+
+        $this->cache->put('flarum-subscriptions.lifetime_stats', $val, static::$lifetimeStatsCacheTtl);
+
+        return $val;
```
Comment (SychO9) [inline extensions/statistics/src/Api/Controller/ShowStatisticsData.php]: can we inject an `\Illuminate\Contracts\Cache\Repository::class` for the cache instead and do this:

```suggestion
        return $this->cache->remember('flarum-subscriptions.lifetime_stats', self::$lifetimeStatsCacheTtl, function () {
            return array_map(function ($entity) {
                return $entity[0]->count();
            }, $this->entities);
        });
```

File: extensions/statistics/src/Api/Controller/ShowStatisticsData.php [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+            return $cachedVal;
+        }
+
+        $val = array_map(function ($entity) {
+            return $this->getTimedCounts($entity[0], $entity[1]);
+        }, $this->entities);
+
+        $this->cache->put('flarum-subscriptions.timed_stats', $val, static::$timedStatsCacheTtl);
+
+        return $val;
```
Comment (SychO9) [inline extensions/statistics/src/Api/Controller/ShowStatisticsData.php]: same here

**Review by SychO9 — CHANGES_REQUESTED** [review]
* For the indexes, do we need both of them, or does only one of them make a difference?
* If we are caching the data I don't think we need to move out the full stats? I'm not entirely sure UX wise if it's less appealing.

File: extensions/statistics/locale/en.yml (hunk 2 above)
Comment (davwheat) [reply extensions/statistics/locale/en.yml:12]: Yeah, I couldn't either

**Comment by davwheat** [comment]
> For the indexes, do we need both of them, or does only one of them make a difference?

Yes, we need both. `type` on its own for the total count, and the composite for the time-based counting. Without both, each can reach up to about 5-6 seconds for 1M models.

**Comment by davwheat** [comment]
For the caching, the reason I still opted to move the full set of data out was the sheer amount of load the statistics gathering could put on the SQL database. I don't think it's necessary to perform multi-second long queries when only loading the admin dashboard, when most of the time the user won't be visiting the page for forum statistics.

**Review by SychO9 — APPROVED** [review]
I always be writing respository as well..

**Pull request merged**